builder/beacon_client.go

```diff
@@ -65,12 +67,10 @@ func (b *BeaconClient) getProposerForNextSlot(requestedSlot uint64) (PubkeyHex,
 }
 
 func (b *BeaconClient) updateValidatorsMap(nextSlot uint64) error {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
 	b.currentSlot = nextSlot
-	nextSlotEpoch := nextSlot / 32
-	if nextSlotEpoch != b.currentEpoch {
+	nextSlotEpoch := nextSlot / b.slotsInEpoch
+	// if slot at half epoch, fetch next epoch's proposers
+	if nextSlotEpoch != b.currentEpoch || nextSlot%b.slotsInEpoch == b.slotsInEpoch/2 {
```

Comment:
Thinking through this, this won't work if there is no validator registered at the mid-epoch slot (won't be called for that slot)